packages/block-editor/src/components/media-replace-flow/test/index.js

```diff
@@ -32,6 +32,34 @@ function TestWrapper() {
 	);
 }
 
+/**
+ * Returns the first found popover element up the DOM tree.
+ *
+ * @param {HTMLElement} element Element to start with.
+ * @return {HTMLElement|null} Popover element, or `null` if not found.
+ */
+function getWrappingPopoverElement( element ) {
+	return element.closest( '.components-popover' );
```

Comment:
Agreed, using `data-test-id` sounds like a better option when possible.

Do we ever add `data-testid` to a component's source, or does that only happen inside the tests ?

Reply:
I'd never recommend cluttering the component source, but for test fixtures, I've found it to be nice and clean as a last resort.

Reply:
I'm afraid I don't get the point of `data-testid` -- on what scale is it a more optimal way to access the popover? That it doesn't trigger the `no-node-access` rule?

Reply:
That, too, but there is more. It also allows us to avoid testing implementation details, and having to peek into the implementation of other components. That makes our tests more resilient since they don't rely on the implementation details of other components we're not currently testing.

Reply:
FWIW I just wanted to point it out as a potential alternative; we can't achieve it here without altering the component source because `popoverProps` don't accept additional props.

Reply:
I think it'll be hard to avoid testing implementation details: the way how the popover is positioned, which element and with what CSS styles, that it's done async -- these are all implementation details. The best we can do is to wrap them inside helpers like custom queries or matchers.

Even the fact that the positioned `div` and the `div` that gets the `data-testid` prop are the same element is an implementation detail. It doesn't always need to be so. The rest props are [internally](https://github.com/WordPress/gutenberg/blob/1906e3923ef138cfd5e6a3984981a425c3d00690/packages/components/src/popover/index.tsx#L193) called `contentProps`, so maybe they were applied to `components-popover_content` element in the past?

Also, as you write, it's not always possible to pass `popoverProps`, especially when testing a larger unit, like a full block edit UI, where there are multiple tooltips and menus.